TFC overrides the default region("global") by "tf-cloud-iso-prod-use1"

### Terraform Version
`Terraform v1.2.4`

### Nomad Version

`Nomad v1.3.1 (2b054e38e91af964d1235faa98c286ca3f527e56)`

### Provider Configuration

```hcl
provider "nomad" {
  address   = var.nomad_address
}

resource "nomad_job" "memcached" {
  jobspec = file("memcached.nomad")
  detach  = false
}
```

`memcached.nomad`:
```hcl
job "memcached" {
  datacenters = ["dc1"]

  group "memcached" {
    task "memcached" {
      driver = "docker"

      config {
        image = "memcached:1.6.15-alpine"
      }
    }
  }
}
```

([See the full code here](https://gitlab.com/lens0021/terraform-nomad-provider-workspace-sample))

### Debug Output
> **Error: error applying jobspec: Unexpected response code: 500 (No path to region)**
with nomad_job.memcached
on base.tf line 29, in resource "nomad_job" "Memcached":
> ```
> resource "nomad_job" "memcached" {
> ```

### Panic Output

> 2022-07-07T10:46:05.769+0900 [WARN]  nomad.rpc: no path found to region: region=tf-cloud-iso-prod-use1
> 2022-07-07T10:46:05.770+0900 [ERROR] http: request failed: method=GET path="/v1/job/memcached?namespace=default&region=tf-cloud-iso-prod-use1" error="No path to region" code=500

### Expected Behavior
I didn't set the region, so the default value `"global"` should be used.

### Actual Behavior
The region seems to be set to `"tf-cloud-iso-prod-use1"`.

### Steps to Reproduce

1. Setup a workspace with the Remote execution mode.
2. Start new run on TFC.
3. Apply.

And as a workaround:

1. Adding `region = "global"` makes the problem disappear.

### References

GH-267 is very similar to this. The problem has shown up without code changes since someday.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,7 +1,8 @@
 ## 1.4.18 (Unreleased)
 
 IMPROVEMENTS:
-* provider: add `load_namespace_env_var` to allow loading the `NOMAD_NAMESPACE` environment variable ([#280](https://github.com/hashicorp/terraform-provider-nomad/pull/280))
+* provider: add `ignore_env_vars` configuration to allow specifying environment variables that should not be loaded by the provider. ([#281](https://github.com/hashicorp/terraform-provider-nomad/pull/281))
+* provider: ignore `NOMAD_NAMESPACE` and `NOMAD_REGION` when running in Terraform Cloud by default. ([#281](https://github.com/hashicorp/terraform-provider-nomad/pull/281))
 
 ## 1.4.17 (June 9, 2022)
 

--- a/nomad/provider.go
+++ b/nomad/provider.go
@@ -33,14 +33,7 @@ func Provider() *schema.Provider {
 			"region": {
 				Type:        schema.TypeString,
 				Optional:    true,
-				DefaultFunc: schema.EnvDefaultFunc("NOMAD_REGION", ""),
 				Description: "Region of the target Nomad agent.",
-			},
-			"load_namespace_env_var": {
-				Type:        schema.TypeBool,
-				Optional:    true,
-				Default:     false,
-				Description: "If true, the NOMAD_NAMESPACE environment variable will be loaded into the provider configuration.",
 			},
 			"http_auth": {
 				Type:        schema.TypeString,
@@ -127,6 +120,12 @@ func Provider() *schema.Provider {
 					},
 				},
 			},
+			"ignore_env_vars": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "A set of environment variables that are ignored by the provider when configuring the Nomad API client.",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
 		},
 
 		ConfigureFunc: providerConfigure,
@@ -179,16 +178,31 @@ func getToken() (string, error) {
 }
 
 func providerConfigure(d *schema.ResourceData) (interface{}, error) {
+	ignoreEnvVars := d.Get("ignore_env_vars").(*schema.Set)
+	if ignoreEnvVars.Len() == 0 {
+		// The Terraform SDK doesn't support DefaultFunc for complex types yet,
+		// so implement the default value logic here for now.
+		// https://github.com/hashicorp/terraform-plugin-sdk/issues/142
+		if os.Getenv("TFC_RUN_ID") != "" {
+			ignoreEnvVars = schema.NewSet(schema.HashString, []interface{}{"NOMAD_NAMESPACE", "NOMAD_REGION"})
+		}
+	}
+
 	conf := api.DefaultConfig()
 	conf.Address = d.Get("address").(string)
-	conf.Region = d.Get("region").(string)
 	conf.SecretID = d.Get("secret_id").(string)
+
+	if region, ok := d.GetOk("region"); ok {
+		conf.Region = region.(string)
+	} else if ignoreEnvVars.Contains("NOMAD_REGION") {
+		conf.Region = ""
+	}
 
 	// The namespace is set per-resource but `DefaultConfig` loads it from the
 	// NOMAD_NAMESPACE env var automatically. This will cause problems when
 	// Terraform is running within a Nomad job (such as in Terraform Cloud) so
 	// we need to unset it unless the provider is configured to load it.
-	if !d.Get("load_namespace_env_var").(bool) {
+	if ignoreEnvVars.Contains("NOMAD_NAMESPACE") {
 		conf.Namespace = ""
 	}
 

--- a/nomad/provider.go
+++ b/nomad/provider.go
@@ -121,10 +121,10 @@ func Provider() *schema.Provider {
 				},
 			},
 			"ignore_env_vars": {
-				Type:        schema.TypeSet,
+				Type:        schema.TypeMap,
 				Optional:    true,
 				Description: "A set of environment variables that are ignored by the provider when configuring the Nomad API client.",
-				Elem:        &schema.Schema{Type: schema.TypeString},
+				Elem:        &schema.Schema{Type: schema.TypeBool},
 			},
 		},
 
@@ -178,13 +178,16 @@ func getToken() (string, error) {
 }
 
 func providerConfigure(d *schema.ResourceData) (interface{}, error) {
-	ignoreEnvVars := d.Get("ignore_env_vars").(*schema.Set)
-	if ignoreEnvVars.Len() == 0 {
+	ignoreEnvVars := d.Get("ignore_env_vars").(map[string]interface{})
+	if len(ignoreEnvVars) == 0 {
 		// The Terraform SDK doesn't support DefaultFunc for complex types yet,
 		// so implement the default value logic here for now.
 		// https://github.com/hashicorp/terraform-plugin-sdk/issues/142
 		if os.Getenv("TFC_RUN_ID") != "" {
-			ignoreEnvVars = schema.NewSet(schema.HashString, []interface{}{"NOMAD_NAMESPACE", "NOMAD_REGION"})
+			ignoreEnvVars = map[string]interface{}{
+				"NOMAD_NAMESPACE": true,
+				"NOMAD_REGION":    true,
+			}
 		}
 	}
 
@@ -194,7 +197,7 @@ func providerConfigure(d *schema.ResourceData) (interface{}, error) {
 
 	if region, ok := d.GetOk("region"); ok {
 		conf.Region = region.(string)
-	} else if ignoreEnvVars.Contains("NOMAD_REGION") {
+	} else if ignore, ok := ignoreEnvVars["NOMAD_REGION"]; ok && ignore.(bool) {
 		conf.Region = ""
 	}
 
@@ -202,7 +205,7 @@ func providerConfigure(d *schema.ResourceData) (interface{}, error) {
 	// NOMAD_NAMESPACE env var automatically. This will cause problems when
 	// Terraform is running within a Nomad job (such as in Terraform Cloud) so
 	// we need to unset it unless the provider is configured to load it.
-	if ignoreEnvVars.Contains("NOMAD_NAMESPACE") {
+	if ignore, ok := ignoreEnvVars["NOMAD_NAMESPACE"]; ok && ignore.(bool) {
 		conf.Namespace = ""
 	}
 

--- a/nomad/provider_test.go
+++ b/nomad/provider_test.go
@@ -168,39 +168,204 @@ func testCheckCSIPluginAvailable(t *testing.T, pluginID string) {
 }
 
 func TestAccNomadProvider_namespace(t *testing.T) {
+	defer func() {
+		os.Unsetenv("NOMAD_NAMESPACE")
+		os.Unsetenv("TFC_RUN_ID")
+	}()
 	var provider *schema.Provider
-	os.Setenv("NOMAD_NAMESPACE", "doesnt-exist")
 
 	resource.Test(t, resource.TestCase{
 		PreCheck:          func() { testAccPreCheck(t) },
 		ProviderFactories: testAccProviderFactoryInternal(&provider),
-		CheckDestroy:      testResourceJob_checkDestroy("provider-namespace-test"),
 		Steps: []resource.TestStep{
+			// Provider reads NOMAD_NAMESPACE by default.
 			{
 				Config: `
-resource "nomad_job" "provider_namespace_test" {
-	jobspec = <<EOT
-		job "provider-namespace-test" {
-			datacenters = ["dc1"]
-			group "foo" {
-				task "foo" {
-					driver = "raw_exec"
-					config {
-						command = "/bin/sleep"
-						args = ["10"]
-					}
-				}
-			}
-		}
-	EOT
-}
+provider "nomad" {}
+
+// necessary to initialize the provider
+data "nomad_namespaces" "test" {}
 				`,
-				// We only need to test that this doesn't fail.
+				PreConfig: func() {
+					os.Setenv("NOMAD_NAMESPACE", "default")
+				},
+				Check: func(_ *terraform.State) error {
+					providerConfig := provider.Meta().(ProviderConfig)
+					config := providerConfig.config
+
+					expect := os.Getenv("NOMAD_NAMESPACE")
+					got := config.Namespace
+					if got != expect {
+						return fmt.Errorf("expected namespace to be %q, got %q", expect, got)
+					}
+					return nil
+				},
+			},
+			// Provider ignores NOMAD_NAMESPACE if requested.
+			{
+				Config: `
+provider "nomad" {
+  ignore_env_vars = ["NOMAD_NAMESPACE"]
+}
+
+// necessary to initialize the provider
+data "nomad_namespaces" "test" {}
+				`,
+				PreConfig: func() {
+					os.Setenv("NOMAD_NAMESPACE", "dev")
+				},
+				Check: func(_ *terraform.State) error {
+					providerConfig := provider.Meta().(ProviderConfig)
+					config := providerConfig.config
+
+					expect := ""
+					got := config.Namespace
+					if got != expect {
+						return fmt.Errorf("expected namespace to be %q, got %q", expect, got)
+					}
+					return nil
+				},
+			},
+			// Provider ignores NOMAD_NAMESPACE if running in TFC.
+			{
+				Config: `
+provider "nomad" {}
+
+// necessary to initialize the provider
+data "nomad_namespaces" "test" {}
+				`,
+				PreConfig: func() {
+					os.Setenv("NOMAD_NAMESPACE", "tfc")
+					os.Setenv("TFC_RUN_ID", "1")
+				},
+				Check: func(_ *terraform.State) error {
+					providerConfig := provider.Meta().(ProviderConfig)
+					config := providerConfig.config
+
+					expect := ""
+					got := config.Namespace
+					if got != expect {
+						return fmt.Errorf("expected namespace to be %q, got %q", expect, got)
+					}
+					return nil
+				},
 			},
 		},
 	})
 
-	os.Unsetenv("NOMAD_NAMESPACE")
+}
+
+func TestAccNomadProvider_region(t *testing.T) {
+	defer func() {
+		os.Unsetenv("NOMAD_REGION")
+		os.Unsetenv("TFC_RUN_ID")
+	}()
+	var provider *schema.Provider
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: testAccProviderFactoryInternal(&provider),
+		Steps: []resource.TestStep{
+			// Provider reads NOMAD_REGION by default.
+			{
+				Config: `
+provider "nomad" {}
+
+// necessary to initialize the provider
+data "nomad_namespaces" "test" {}
+				`,
+				PreConfig: func() {
+					os.Setenv("NOMAD_REGION", "global")
+				},
+				Check: func(_ *terraform.State) error {
+					providerConfig := provider.Meta().(ProviderConfig)
+					config := providerConfig.config
+
+					expect := os.Getenv("NOMAD_REGION")
+					got := config.Region
+					if got != expect {
+						return fmt.Errorf("expected region to be %q, got %q", expect, got)
+					}
+					return nil
+				},
+			},
+			// Provider overrides NOMAD_REGION if specified in the config.
+			{
+				Config: `
+provider "nomad" {
+  region = "global"
+}
+
+// necessary to initialize the provider
+data "nomad_namespaces" "test" {}
+				`,
+				PreConfig: func() {
+					os.Setenv("NOMAD_REGION", "us")
+				},
+				Check: func(_ *terraform.State) error {
+					providerConfig := provider.Meta().(ProviderConfig)
+					config := providerConfig.config
+
+					expect := "global"
+					got := config.Region
+					if got != expect {
+						return fmt.Errorf("expected region to be %q, got %q", expect, got)
+					}
+					return nil
+				},
+			},
+			// Provider ignores NOMAD_REGION if requested.
+			{
+				Config: `
+provider "nomad" {
+  ignore_env_vars = ["NOMAD_REGION"]
+}
+
+// necessary to initialize the provider
+data "nomad_namespaces" "test" {}
+				`,
+				PreConfig: func() {
+					os.Setenv("NOMAD_REGION", "us")
+				},
+				Check: func(_ *terraform.State) error {
+					providerConfig := provider.Meta().(ProviderConfig)
+					config := providerConfig.config
+
+					expect := ""
+					got := config.Region
+					if got != expect {
+						return fmt.Errorf("expected region to be %q, got %q", expect, got)
+					}
+					return nil
+				},
+			},
+			// Provider ignores NOMAD_REGION if running in TFC.
+			{
+				Config: `
+provider "nomad" {}
+
+// necessary to initialize the provider
+data "nomad_namespaces" "test" {}
+				`,
+				PreConfig: func() {
+					os.Setenv("NOMAD_REGION", "tfc")
+					os.Setenv("TFC_RUN_ID", "1")
+				},
+				Check: func(_ *terraform.State) error {
+					providerConfig := provider.Meta().(ProviderConfig)
+					config := providerConfig.config
+
+					expect := ""
+					got := config.Region
+					if got != expect {
+						return fmt.Errorf("expected region to be %q, got %q", expect, got)
+					}
+					return nil
+				},
+			},
+		},
+	})
+
 }
 
 func TestAccNomadProvider_Headers(t *testing.T) {

--- a/nomad/provider_test.go
+++ b/nomad/provider_test.go
@@ -205,7 +205,9 @@ data "nomad_namespaces" "test" {}
 			{
 				Config: `
 provider "nomad" {
-  ignore_env_vars = ["NOMAD_NAMESPACE"]
+  ignore_env_vars = {
+    NOMAD_NAMESPACE: true,
+  }
 }
 
 // necessary to initialize the provider
@@ -243,6 +245,34 @@ data "nomad_namespaces" "test" {}
 					config := providerConfig.config
 
 					expect := ""
+					got := config.Namespace
+					if got != expect {
+						return fmt.Errorf("expected namespace to be %q, got %q", expect, got)
+					}
+					return nil
+				},
+			},
+			// Provider loads NOMAD_NAMESPACE even when running in TFC if user requests.
+			{
+				Config: `
+provider "nomad" {
+  ignore_env_vars = {
+    NOMAD_NAMESPACE: false,
+  }
+}
+
+// necessary to initialize the provider
+data "nomad_namespaces" "test" {}
+				`,
+				PreConfig: func() {
+					os.Setenv("NOMAD_NAMESPACE", "default")
+					os.Setenv("TFC_RUN_ID", "1")
+				},
+				Check: func(_ *terraform.State) error {
+					providerConfig := provider.Meta().(ProviderConfig)
+					config := providerConfig.config
+
+					expect := os.Getenv("NOMAD_NAMESPACE")
 					got := config.Namespace
 					if got != expect {
 						return fmt.Errorf("expected namespace to be %q, got %q", expect, got)
@@ -318,7 +348,9 @@ data "nomad_namespaces" "test" {}
 			{
 				Config: `
 provider "nomad" {
-  ignore_env_vars = ["NOMAD_REGION"]
+  ignore_env_vars = {
+    NOMAD_REGION: true,
+  }
 }
 
 // necessary to initialize the provider
@@ -356,6 +388,34 @@ data "nomad_namespaces" "test" {}
 					config := providerConfig.config
 
 					expect := ""
+					got := config.Region
+					if got != expect {
+						return fmt.Errorf("expected region to be %q, got %q", expect, got)
+					}
+					return nil
+				},
+			},
+			// Provider loads NOMAD_REGION even when running in TFC if user requests.
+			{
+				Config: `
+provider "nomad" {
+  ignore_env_vars = {
+    NOMAD_REGION: false,
+  }
+}
+
+// necessary to initialize the provider
+data "nomad_namespaces" "test" {}
+				`,
+				PreConfig: func() {
+					os.Setenv("NOMAD_REGION", "global")
+					os.Setenv("TFC_RUN_ID", "1")
+				},
+				Check: func(_ *terraform.State) error {
+					providerConfig := provider.Meta().(ProviderConfig)
+					config := providerConfig.config
+
+					expect := os.Getenv("NOMAD_REGION")
 					got := config.Region
 					if got != expect {
 						return fmt.Errorf("expected region to be %q, got %q", expect, got)

--- a/website/docs/index.html.markdown
+++ b/website/docs/index.html.markdown
@@ -40,13 +40,6 @@ The following arguments are supported:
 - `region` `(string: "")` - The Nomad region to target. This can also be
   specified as the `NOMAD_REGION` environment variable.
 
-- `load_namespace_env_var` `(bool: false)` - If true, the `NOMAD_NAMESPACE`
-  environment variable will be loaded into the provider configuration.
-
-    ~> **Warning:** This value should not be set (or set to `false`) when
-      running Terraform in environments where it runs within a Nomad
-      allocation, such as in Terraform Cloud.
-
 - `http_auth` `(string: "")` - HTTP Basic Authentication credentials to be used
   when communicating with Nomad, in the format of either `user` or `user:pass`.
   This can also be specified using the `NOMAD_HTTP_AUTH` environment variable.
@@ -90,6 +83,12 @@ The following arguments are supported:
 - `secret_id` `(string: "")` - The Secret ID of an ACL token to make requests with,
   for ACL-enabled clusters. This can also be specified via the `NOMAD_TOKEN`
   environment variable.
+
+- `ignore_env_vars` `([]string: [])` - A set of environment variables that are
+  ignored by the provider when configuring the Nomad API client. Supported
+  values are: `NOMAD_NAMESPACE` and `NOMAD_REGION`. When using the provider
+  with Terraform Cloud, the default value is set to `["NOMAD_NAMESPACE",
+  "NOMAD_REGION"]`.
 
 The `headers` configuration block accepts the following arguments:
 * `name` - (Required) The name of the header.

--- a/website/docs/index.html.markdown
+++ b/website/docs/index.html.markdown
@@ -84,11 +84,17 @@ The following arguments are supported:
   for ACL-enabled clusters. This can also be specified via the `NOMAD_TOKEN`
   environment variable.
 
-- `ignore_env_vars` `([]string: [])` - A set of environment variables that are
-  ignored by the provider when configuring the Nomad API client. Supported
-  values are: `NOMAD_NAMESPACE` and `NOMAD_REGION`. When using the provider
-  with Terraform Cloud, the default value is set to `["NOMAD_NAMESPACE",
-  "NOMAD_REGION"]`.
+- `ignore_env_vars` `(map[string]bool: {})` - A map of environment variables
+  that are ignored by the provider when configuring the Nomad API client.
+  Supported keys are: `NOMAD_NAMESPACE` and `NOMAD_REGION`. When using the
+  provider within Terraform Cloud, the default value is set to
+    ```
+    {
+      NOMAD_NAMESPACE: true,
+      NOMAD_REGION:    true,
+    }
+    ```.
+  Set these values to `false` if you need to load these environment variables.
 
 The `headers` configuration block accepts the following arguments:
 * `name` - (Required) The name of the header.

